api/server/middleware_auth.go

```diff
@@ -38,6 +38,76 @@ func NewAuthMiddleware() *authMiddleware {
 type authMiddleware struct {
 }
 
+// newSecurityMiddleware based on auth configuration returns securityHandler or just
+func newSecurityMiddleware(authenticators map[string]auth.Authenticator) func(next http.HandlerFunc) http.HandlerFunc {
+	if auth.Enabled() {
+		return func(next http.HandlerFunc) http.HandlerFunc {
+			return securityHandler(authenticators, next)
+		}
+	}
+
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return next
+	}
+}
+
+// securityHandler implements Authentication and Authorization check at the same time
+// this functionality where not moved to separate functions because of simplicity
+func securityHandler(authenticators map[string]auth.Authenticator, next http.HandlerFunc) http.HandlerFunc {
+	if authenticators == nil {
+		return next
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		tokenHeader := r.Header.Get("Authorization")
+		tokens := strings.Split(tokenHeader, " ")
+
+		if len(tokens) < 2 {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		token := tokens[1]
+
+		// Determine issuer
+		issuer, err := auth.TokenIssuer(token)
+		if err != nil {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+
+		// Authenticate user
+		ctx := r.Context()
+		var claims *auth.Claims
+		if authenticator, exists := authenticators[issuer]; exists {
+			claims, err = authenticator.AuthenticateToken(ctx, token)
+			if err != nil || claims == nil {
+				w.WriteHeader(http.StatusForbidden)
+				return
+			}
+		} else {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+
+		isSystemAdmin := false
+
+		for _, role := range claims.Roles {
+			if role == "system.admin" {
+				isSystemAdmin = true
+				break
+			}
+		}
+
+		if !isSystemAdmin {
+			w.WriteHeader(http.StatusUnauthorized)
```

Comment:
I think, I may be wrong, that you can call `w.Write()` first to write to the return body. You can write that only `system.admin` role is allowed.

Reply:
Well, I use only WriteHeader method of ResponseWriter in that method, Write can be call only in downstream handlers.